### ISSUE TITLE
400 error on Settings view for a New User

### ISSUE THREAD

**Issue by mk4111** [emfoundation/ce100-app#558]
'Settings' view returns a 400 error when we create a new primary user (for a new organisation), activate the account (and therefore add a password), and visit Settings in the top right corner from the new primary user's account. The url path seems to be correct at `/people/13/edit` though.

This issue does not repeat for newly created Admins.

![screenshot from 2017-01-12 16-04-34](https://cloud.githubusercontent.com/assets/13470325/21897274/02298598-d8e1-11e6-985f-8d2063d117c8.png)

**Comment by JMurphyWeb** [comment]
Currently getting a problem that I cannot add a new user without getting  500 error. could be a send email issue.

**Comment by SimonLab** [comment]
@JMurphyWeb I saw your [PR](https://github.com/dwyl/hapi-error/pull/44) on hapi-error to display the stack error has been merged :tada: So it might be good to display the errors when the status code is 500 it will be much easier to target the issues!

**Comment by JMurphyWeb** [comment]
I am currently unable to create a new user so this feature is impossible to debug: https://github.com/emfoundation/ce100-app/issues/572

I need to check why hapi error is not updated locally (and therefore not printing the error stack trace)

**Comment by mk4111** [comment]
I've been able to create a new user so have taken this issue over from @JMurphyWeb.

The page breaks following the below userflow:
- Add Org
- Add New User
- Activate New User + Add New Password
- Go to Settings (as new primary user)

The url at which it breaks is: `/people/12/edit`

Looking at the database we see that the user does exist, and that the new user has id = 12.

![screenshot from 2017-01-13 14-42-35](https://cloud.githubusercontent.com/assets/13470325/21933758/8c802d74-d99e-11e6-8f91-0ced9fcd034d.png)

**Comment by mk4111** [comment]
Following the above, visiting '/login' and logging back into the app allows the user to view User Settings.

**Comment by mk4111** [comment]
Having placed a console.log in the callback following the `getBy` method in edit-view.js handler, we see that nothing is returned.

![screenshot from 2017-01-13 15-01-17](https://cloud.githubusercontent.com/assets/13470325/21934398/2b798fae-d9a1-11e6-94db-79f494069481.png)

This suggests that our `getBy` query might be using `last_login` as a requirement/condition in the query to postgres

**Comment by mk4111** [comment]
Sorry! Late update! 

So it seems the issue was occurring _before_ the `getBy` function where we would catch whether `loggedIn.userId !== editId` where `editId = request.params.id && JSON.parse(request.params.id)`.

I wondered if it was the conditional `&&` that was the cause for the break, whereby JSON.parse(editId) failed and so the value reverted to the first statement, but this doesn't quite make sense because editId is clearly returned as `12`...

This also doesn't quite make sense as, once the user is logged in, the same catch no longer applies, so loggedIn.userId === editId.

**Comment by mk4111** [comment]
making `loggedIn.userId !== editId` --> `JSON.parse(loggedIn.userId !== editId)` seems to do the trick.

**Comment by SimonLab** [comment]
I don't think it's a good idea to use JSON.parse to convert a String into an Integer, see my [comment on the PR](https://github.com/emfoundation/ce100-app/pull/574#pullrequestreview-16721496).

@mk4111 I think you manage to find where and why the code was failing :+1: We just need to understand now why the error was appearing and why the type is not Integer as we were expected, like this we are sure that this issue doesn't hide another issue:

When an admin creates a new user the handler is creating a jwt string which contains only the id of the new user created, see [this line](https://github.com/emfoundation/ce100-app/blob/e718b4e50ac338df380c25f5cd2f92953f154ca4/server/handlers/people/add.js#L25) at this step the id passed to jwt.sign is indeed an Integer. However jwt.sign as the following signature:
```
jwt.sign(payload, secretOrPrivateKey, options, [callback])
```
with this comment on its documentation:

```
If payload is not a buffer or a string, it will be coerced into a string using JSON.stringify.
```

So here our payload is the id of the new created user which is an Integer. So jwt.sign will first convert this id into a String then create the jwt string token. This token is then send by email to the new created user which will be verify on the activate handler, see [this line](https://github.com/emfoundation/ce100-app/blob/e718b4e50ac338df380c25f5cd2f92953f154ca4/server/handlers/people/activate-account.js#L15). So at this stage the decoded userId from the token is a string which is then used for the new session of the user.

So to recap the issue is due to the way jwt.sign convert the Integer payload to a string. There are no other issues at the origin of this one so we can easily fix it by using parseInt to ensure that the type of the id is an Integer

**Comment by mk4111** [comment]
@SimonLab - this is exactly why we need you <3 <3 <3 Thank you!!!!! <3
